pyro/infer/reparam/split.py

```diff
@@ -0,0 +1,90 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from torch.distributions import constraints
+
+import pyro
+import pyro.distributions as dist
+from pyro.distributions.util import broadcast_shape
+
+from .reparam import Reparam
+
+
+class _ImproperUniform(dist.TorchDistribution):
+    """
+    Internal helper distribution with zero :meth:`log_prob` and undefined
+    :meth:`sample`.
+    """
+    arg_constraints = {}
+
+    def __init__(self, support, batch_shape, event_shape):
+        self._support = support
+        super().__init__(batch_shape, event_shape)
+
+    @constraints.dependent_property
+    def support(self):
+        return self._support
+
+    def expand(self, batch_shape, _instance=None):
+        batch_shape = torch.Size(batch_shape)
+        new = self._get_checked_instance(_ImproperUniform, _instance)
+        new._support = self._support
+        super(_ImproperUniform, new).__init__(batch_shape, self.event_shape)
+        return new
+
+    def log_prob(self, value):
+        batch_shape = value.shape[:value.dim() - self.event_dim]
+        batch_shape = broadcast_shape(batch_shape, self.batch_shape)
+        return torch.zeros(()).expand(batch_shape)
+
+    def sample(self, sample_shape=torch.Size()):
+        raise NotImplementedError("SplitReparam does not support sampling")
```

Comment:
nit: `_ImproperUniform` does not support sampling

Reply:
> nit: _ImproperUniform does not support sampling

This user-facing exception mention the user-facing class; `_ImproperUniform` is an implementation detail.

> I think we can move this to the main distributions module.

I feel like this is too bespoke for general use. Until we find another use case I'd prefer to keep it private. We may find another way to implement this, e.g. `Delta(nan).mask(False)` would raise a Nan error rather than a `NotImplementedError`.

Reply:
> with an sample_fn arg to generate prototype samples for HMC.

I believe this functionality is already cleanly accomplished by `InitMessenger`. Do you have another use case in mind?

Reply:
Nevermind, I think `dist.Foo(...).mask(False)` can serve the same purpose and initial values can be drawn directly from `Foo` distribution. This class is useful when we don't know the support before-hand.